### ISSUE TITLE
Verify whether we have global public symbols in the upcoming 0.12.0 that are not part of the public API

### ISSUE THREAD

**Issue by andy-maier** [pywbem/pywbem#850]
We state that the public API of pywbem is what is in the pywbem namespace. This is the sum of what is in the "all" arrays in the sub-modules.

There may be additional symbols in the submodules that do not have a leadnig underscore, so people may have used them.

This issue is to make a list of them, and to discuss what we do with them.

Here is a command to list all the non-underscore classes and functions:
```
grep -E "^(class |def )[^_]" pywbem/*.py |cut -d ' ' -f 2 |cut -d '(' -f 1 |sort |uniq
```

**Comment by andy-maier** [comment]
There is an additional issue, namely whether all symbols in the pywbem namespace are actually documented.

Here is the answer to both:

Symbols that should probably be public but are missing in the pywbem namespace and in the documentation:

```
_logging.py               var   LOG_HTTP_NAME
_logging.py               var   DEFAULT_LOG_DETAIL_LEVEL
```

Symbols with public names in submodules with public names, that are not in the pywbem namespace:
```
cim_constants.py          var   PROVIDERTYPE_ASSOCIATION
cim_constants.py          var   PROVIDERTYPE_CLASS
cim_constants.py          var   PROVIDERTYPE_CONSUMER
cim_constants.py          var   PROVIDERTYPE_INDICATION
cim_constants.py          var   PROVIDERTYPE_INSTANCE
cim_constants.py          var   PROVIDERTYPE_METHOD
cim_constants.py          var   PROVIDERTYPE_QUERY

cim_http.py               var   DEFAULT_PORT_HTTP
cim_http.py               var   DEFAULT_PORT_HTTPS
cim_http.py               func  create_pywbem_ssl_context
cim_http.py               func  get_default_ca_cert_paths
cim_http.py               func  get_default_ca_certs
cim_http.py               func  get_object_header
cim_http.py               func  parse_url
cim_http.py               func  wbem_request
cim_http.py               class HTTPTimeout

cim_obj.py                var   MAX_MOF_LINE
cim_obj.py                var   MOF_INDENT
cim_obj.py                func  byname
cim_obj.py                func  check_array_parms
cim_obj.py                func  check_embedded_object
cim_obj.py                func  cmpitem
cim_obj.py                func  cmpname
cim_obj.py                func  infer_embedded_object
cim_obj.py                func  infer_is_array
cim_obj.py                func  infer_type
cim_obj.py                func  mofstr
cim_obj.py                func  moftype
cim_obj.py                func  partition

cim_operations.py         var   pull_inst_result_tuple
cim_operations.py         var   pull_path_result_tuple
cim_operations.py         var   pull_query_result_tuple
cim_operations.py         func  check_utf8_xml_chars
cim_operations.py         func  is_subclass
cim_operations.py         func  OpenWBEMUDSConnection
cim_operations.py         func  PegasusUDSConnection
cim_operations.py         func  SFCBUDSConnection

cim_types.py              func  atomic_to_cim_xml
cim_types.py              func  cimtype
cim_types.py              func  type_from_name

cim_xml.py                class CIM
cim_xml.py                class CIMElement
cim_xml.py                class CLASS
cim_xml.py                class CLASSNAME
cim_xml.py                class CLASSPATH
cim_xml.py                class DECLARATION
cim_xml.py                class DECLGROUP
cim_xml.py                class DECLGROUP_WITHNAME
cim_xml.py                class DECLGROUP_WITHPATH
cim_xml.py                class ERROR
cim_xml.py                class EXPMETHODCALL
cim_xml.py                class EXPMETHODRESPONSE
cim_xml.py                class EXPPARAMVALUE
cim_xml.py                class HOST
cim_xml.py                class IMETHODCALL
cim_xml.py                class IMETHODRESPONSE
cim_xml.py                class INSTANCE
cim_xml.py                class INSTANCENAME
cim_xml.py                class INSTANCEPATH
cim_xml.py                class IPARAMVALUE
cim_xml.py                class IRETURNVALUE
cim_xml.py                class KEYBINDING
cim_xml.py                class KEYVALUE
cim_xml.py                class LOCALCLASSPATH
cim_xml.py                class LOCALINSTANCEPATH
cim_xml.py                class LOCALNAMESPACEPATH
cim_xml.py                class MESSAGE
cim_xml.py                class METHOD
cim_xml.py                class METHODCALL
cim_xml.py                class METHODRESPONSE
cim_xml.py                class MULTIEXPREQ
cim_xml.py                class MULTIEXPRSP
cim_xml.py                class MULTIREQ
cim_xml.py                class MULTIRSP
cim_xml.py                class NAMESPACE
cim_xml.py                class NAMESPACEPATH
cim_xml.py                class OBJECTPATH
cim_xml.py                class PARAMETER
cim_xml.py                class PARAMETER_ARRAY
cim_xml.py                class PARAMETER_REFARRAY
cim_xml.py                class PARAMETER_REFERENCE
cim_xml.py                class PARAMVALUE
cim_xml.py                class PROPERTY
cim_xml.py                class PROPERTY_ARRAY
cim_xml.py                class PROPERTY_REFERENCE
cim_xml.py                class QUALIFIER
cim_xml.py                class QUALIFIER_DECLARATION
cim_xml.py                class RESPONSEDESTINATION
cim_xml.py                class RETURNVALUE
cim_xml.py                class SCOPE
cim_xml.py                class SIMPLEEXPREQ
cim_xml.py                class SIMPLEEXPRSP
cim_xml.py                class SIMPLEREQ
cim_xml.py                class SIMPLEREQACK
cim_xml.py                class SIMPLERSP
cim_xml.py                class TABLE
cim_xml.py                class TABLECELL_DECLARATION
cim_xml.py                class TABLECELL_REFERENCE
cim_xml.py                class TABLEROW
cim_xml.py                class TABLEROW_DECLARATION
cim_xml.py                class VALUE
cim_xml.py                class VALUE_ARRAY
cim_xml.py                class VALUE_NAMEDINSTANCE
cim_xml.py                class VALUE_NAMEDOBJECT
cim_xml.py                class VALUE_NULL
cim_xml.py                class VALUE_OBJECT
cim_xml.py                class VALUE_OBJECTWITHLOCALPATH
cim_xml.py                class VALUE_OBJECTWITHPATH
cim_xml.py                class VALUE_REFARRAY
cim_xml.py                class VALUE_REFERENCE

mof_compiler.py           var   cChar
mof_compiler.py           var   charvalue_re
mof_compiler.py           var   escapeSequence
mof_compiler.py           var   hexEscape
mof_compiler.py           var   identifier_re
mof_compiler.py           var   literals
mof_compiler.py           var   reserved
mof_compiler.py           var   sChar
mof_compiler.py           var   simpleEscape
mof_compiler.py           var   stringvalue_re
mof_compiler.py           var   t_ignore
mof_compiler.py           var   tokens
mof_compiler.py           var   utf8_2
mof_compiler.py           var   utf8_3_1
mof_compiler.py           var   utf8_3_2
mof_compiler.py           var   utf8_3_3
mof_compiler.py           var   utf8_3_4
mof_compiler.py           var   utf8_4_1
mof_compiler.py           var   utf8_4_2
mof_compiler.py           var   utf8_4_3
mof_compiler.py           var   utf8Char
mof_compiler.py           func  p_alias
mof_compiler.py           func  p_aliasIdentifier
mof_compiler.py           func  p_array
mof_compiler.py           func  p_arrayInitializer
mof_compiler.py           func  p_assocDeclaration
mof_compiler.py           func  p_associationFeature
mof_compiler.py           func  p_associationFeatureList
mof_compiler.py           func  p_booleanValue
mof_compiler.py           func  p_classDeclaration
mof_compiler.py           func  p_classFeature
mof_compiler.py           func  p_classFeatureList
mof_compiler.py           func  p_className
mof_compiler.py           func  p_compilerDirective
mof_compiler.py           func  p_constantValue
mof_compiler.py           func  p_constantValueList
mof_compiler.py           func  p_dataType
mof_compiler.py           func  p_defaultFlavor
mof_compiler.py           func  p_defaultValue
mof_compiler.py           func  p_empty
mof_compiler.py           func  p_error
mof_compiler.py           func  p_flavor
mof_compiler.py           func  p_flavorList
mof_compiler.py           func  p_flavorListWithComma
mof_compiler.py           func  p_identifier
mof_compiler.py           func  p_indicDeclaration
mof_compiler.py           func  p_initializer
mof_compiler.py           func  p_instanceDeclaration
mof_compiler.py           func  p_integerValue
mof_compiler.py           func  p_metaElement
mof_compiler.py           func  p_metaElementList
mof_compiler.py           func  p_methodDeclaration
mof_compiler.py           func  p_methodName
mof_compiler.py           func  p_mofProduction
mof_compiler.py           func  p_mofProductionList
mof_compiler.py           func  p_mofSpecification
mof_compiler.py           func  p_mp_createClass
mof_compiler.py           func  p_mp_createInstance
mof_compiler.py           func  p_mp_setQualifier
mof_compiler.py           func  p_nullValue
mof_compiler.py           func  p_objectHandle
mof_compiler.py           func  p_objectRef
mof_compiler.py           func  p_parameter
mof_compiler.py           func  p_parameter_1
mof_compiler.py           func  p_parameter_2
mof_compiler.py           func  p_parameter_3
mof_compiler.py           func  p_parameter_4
mof_compiler.py           func  p_parameterList
mof_compiler.py           func  p_parameterName
mof_compiler.py           func  p_pragmaName
mof_compiler.py           func  p_pragmaParameter
mof_compiler.py           func  p_propertyDeclaration
mof_compiler.py           func  p_propertyDeclaration_1
mof_compiler.py           func  p_propertyDeclaration_2
mof_compiler.py           func  p_propertyDeclaration_3
mof_compiler.py           func  p_propertyDeclaration_4
mof_compiler.py           func  p_propertyDeclaration_5
mof_compiler.py           func  p_propertyDeclaration_6
mof_compiler.py           func  p_propertyDeclaration_7
mof_compiler.py           func  p_propertyDeclaration_8
mof_compiler.py           func  p_propertyName
mof_compiler.py           func  p_qualifier
mof_compiler.py           func  p_qualifierDeclaration
mof_compiler.py           func  p_qualifierList
mof_compiler.py           func  p_qualifierListEmpty
mof_compiler.py           func  p_qualifierName
mof_compiler.py           func  p_qualifierParameter
mof_compiler.py           func  p_qualifierType
mof_compiler.py           func  p_qualifierType_1
mof_compiler.py           func  p_qualifierType_2
mof_compiler.py           func  p_referenceDeclaration
mof_compiler.py           func  p_referenceInitializer
mof_compiler.py           func  p_referenceName
mof_compiler.py           func  p_scope
mof_compiler.py           func  p_stringValueList
mof_compiler.py           func  p_superClass
mof_compiler.py           func  p_valueInitializer
mof_compiler.py           func  p_valueInitializerList
mof_compiler.py           func  t_binaryValue
mof_compiler.py           func  t_charValue
mof_compiler.py           func  t_COMMENT
mof_compiler.py           func  t_decimalValue
mof_compiler.py           func  t_error
mof_compiler.py           func  t_floatValue
mof_compiler.py           func  t_hexValue
mof_compiler.py           func  t_IDENTIFIER
mof_compiler.py           func  t_MCOMMENT
mof_compiler.py           func  t_newline
mof_compiler.py           func  t_octalValue
mof_compiler.py           func  t_stringValue

tupleparse.py             func  attrs
tupleparse.py             func  check_node
tupleparse.py             func  filter_tuples
tupleparse.py             func  kids
tupleparse.py             func  list_of_matching
tupleparse.py             func  list_of_same
tupleparse.py             func  list_of_various
tupleparse.py             func  name
tupleparse.py             func  notimplemented
tupleparse.py             func  one_child
tupleparse.py             func  optional_child
tupleparse.py             func  parse_any
tupleparse.py             func  parse_cim
tupleparse.py             func  parse_class
tupleparse.py             func  parse_classname
tupleparse.py             func  parse_classpath
tupleparse.py             func  parse_declaration
tupleparse.py             func  parse_declgroup
tupleparse.py             func  parse_embeddedObject
tupleparse.py             func  parse_error
tupleparse.py             func  parse_expmethodcall
tupleparse.py             func  parse_expmethodresponse
tupleparse.py             func  parse_expparamvalue
tupleparse.py             func  parse_host
tupleparse.py             func  parse_imethodcall
tupleparse.py             func  parse_imethodresponse
tupleparse.py             func  parse_instance
tupleparse.py             func  parse_instancename
tupleparse.py             func  parse_instancepath
tupleparse.py             func  parse_iparamvalue
tupleparse.py             func  parse_ireturnvalue
tupleparse.py             func  parse_keybinding
tupleparse.py             func  parse_keyvalue
tupleparse.py             func  parse_localclasspath
tupleparse.py             func  parse_localinstancepath
tupleparse.py             func  parse_localnamespacepath
tupleparse.py             func  parse_message
tupleparse.py             func  parse_method
tupleparse.py             func  parse_methodcall
tupleparse.py             func  parse_methodresponse
tupleparse.py             func  parse_multiexpreq
tupleparse.py             func  parse_multiexprsp
tupleparse.py             func  parse_multireq
tupleparse.py             func  parse_multirsp
tupleparse.py             func  parse_namespace
tupleparse.py             func  parse_namespacepath
tupleparse.py             func  parse_objectpath
tupleparse.py             func  parse_parameter
tupleparse.py             func  parse_parameter_array
tupleparse.py             func  parse_parameter_refarray
tupleparse.py             func  parse_parameter_reference
tupleparse.py             func  parse_paramvalue
tupleparse.py             func  parse_property
tupleparse.py             func  parse_property_array
tupleparse.py             func  parse_property_reference
tupleparse.py             func  parse_qualifier
tupleparse.py             func  parse_qualifier_declaration
tupleparse.py             func  parse_returnvalue
tupleparse.py             func  parse_scope
tupleparse.py             func  parse_simpleexpreq
tupleparse.py             func  parse_simpleexprsp
tupleparse.py             func  parse_simplereq
tupleparse.py             func  parse_simplersp
tupleparse.py             func  parse_value
tupleparse.py             func  parse_value_array
tupleparse.py             func  parse_value_instancewithpath
tupleparse.py             func  parse_value_namedinstance
tupleparse.py             func  parse_value_namedobject
tupleparse.py             func  parse_value_object
tupleparse.py             func  parse_value_objectwithlocalpath
tupleparse.py             func  parse_value_objectwithpath
tupleparse.py             func  parse_value_refarray
tupleparse.py             func  parse_value_reference
tupleparse.py             func  pcdata
tupleparse.py             func  unpack_boolean
tupleparse.py             func  unpack_value

tupletree.py              func  xml_to_tupletree_sax
tupletree.py              class CIMContentHandler
```

**Comment by andy-maier** [comment]
**Proposal to be discussed:**

1. Make the following symbols that were probably missed, public in pywbem namespace and documentation:
  ```
_logging.py               var   LOG_HTTP_NAME
_logging.py               var   DEFAULT_LOG_DETAIL_LEVEL
```

2. Change the names of the following symbols that were introduced in 0.12.0 to become private in the submodule by adding a leading underscore:
```
cim_obj.py                func  check_array_parms                0.12.0
cim_obj.py                func  check_embedded_object            0.12.0
cim_obj.py                func  infer_embedded_object            0.12.0
cim_obj.py                func  infer_is_array                   0.12.0
cim_obj.py                func  infer_type                       0.12.0
cim_obj.py                func  partition                        0.12.0
```

3. Leave the names of public symbols in submodules that were introduced between 0.8 and 0.11 unchanged. The submodule namespaces can be removed from the pywbem namespace in 1.0.0, which will take care of these symbols by then, and it is not really necessary to introduce a potential incompatibility now.

**Comment by andy-maier** [comment]
We accepted the proposal and go forward with it.

**Comment by andy-maier** [comment]
The proposal is addressed in PR #885.
